### ISSUE TITLE
Linux Vault binary panics after a successful LDAP login

### ISSUE THREAD

**Issue by hobbeswalsh** [hashicorp/vault#672]
Let's try logging in with the wrong LDAP password:

```
$ /vault auth -method=ldap username="My Name"
Password (will be hidden):  <enter wrong password here>
Error making API request.

URL: PUT https://vault.counsyl.com/v1/auth/ldap/login/My%!N(MISSING)ame
Code: 400. Errors:

* LDAP bind failed: LDAP Result Code 49 "Invalid Credentials":
```

Ok fine, that looks like an expected failure. Now let's try logging in with the _right_ LDAP password.

```
$ ./vault auth -method=ldap username="My Name"
Password (will be hidden): <enter the right password>
panic: interface conversion: interface is nil, not float64

goroutine 1 [running]:
github.com/hashicorp/vault/command.(*AuthCommand).Run(0xc8200fd810, 0xc82000a1b0, 0x1, 0x1, 0xc820142098)
	/gopath/src/github.com/hashicorp/vault/command/auth.go:206 +0x2041
github.com/mitchellh/cli.(*CLI).Run(0xc8200fd760, 0xc8200fd760, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/Godeps/_workspace/src/github.com/mitchellh/cli/cli.go:112 +0x655
github.com/hashicorp/vault/cli.RunCustom(0xc82000a190, 0x3, 0x3, 0xc820128720, 0x2)
	/gopath/src/github.com/hashicorp/vault/cli/main.go:44 +0x452
github.com/hashicorp/vault/cli.Run(0xc82000a190, 0x3, 0x3, 0x989300)
	/gopath/src/github.com/hashicorp/vault/cli/main.go:11 +0x4c
main.main()
	/gopath/src/github.com/hashicorp/vault/main.go:10 +0x60

goroutine 5 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 7 [semacquire]:
sync.runtime_Syncsemacquire(0xc820146040)
	/usr/local/go/src/runtime/sema.go:237 +0x201
sync.(*Cond).Wait(0xc820146030)
	/usr/local/go/src/sync/cond.go:62 +0x9b
io.(*pipe).read(0xc820146000, 0xc820168000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:52 +0x2d2
io.(*PipeReader).Read(0xc820028168, 0xc820168000, 0x1000, 0x1000, 0xc82001a418, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:134 +0x50
bufio.(*Scanner).Scan(0xc820016800, 0x2)
	/usr/local/go/src/bufio/scan.go:180 +0x877
github.com/hashicorp/vault/command.(*Meta).FlagSet.func1(0xc820016800, 0xc8200fd810)
	/gopath/src/github.com/hashicorp/vault/command/meta.go:239 +0x25
created by github.com/hashicorp/vault/command.(*Meta).FlagSet
	/gopath/src/github.com/hashicorp/vault/command/meta.go:242 +0x5e6

goroutine 10 [select, locked to thread]:
runtime.gopark(0x1380530, 0xc820026f28, 0x1121188, 0x6, 0x42c518, 0x2)
	/usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc820026f28, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc820026f28)
	/usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 18 [IO wait]:
net.runtime_pollWait(0x2b1a1329d648, 0x72, 0xc82000e160)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82012de20, 0x72, 0x0, 0x0)
	/goroot/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82012de20, 0x0, 0x0)
	/goroot/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82012ddc0, 0xc8200d1400, 0x400, 0x400, 0x0, 0x2b1a13297050, 0xc82000e160)
	/goroot/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820028238, 0xc8200d1400, 0x400, 0x400, 0x0, 0x0, 0x0)
	/goroot/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc8201921b0, 0x2b1a1329da98, 0xc820028238, 0x5, 0x0, 0x0)
	/goroot/src/crypto/tls/conn.go:455 +0xcc
crypto/tls.(*Conn).readRecord(0xc82007b340, 0x1380517, 0x0, 0x0)
	/goroot/src/crypto/tls/conn.go:540 +0x2d1
crypto/tls.(*Conn).Read(0xc82007b340, 0xc82018b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/goroot/src/crypto/tls/conn.go:901 +0x167
net/http.noteEOFReader.Read(0x2b1a132a62d8, 0xc82007b340, 0xc8200fda78, 0xc82018b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/goroot/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc820480800, 0xc82018b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc8203cf980)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8203cf980, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8200fda20)
	/goroot/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/goroot/src/net/http/transport.go:685 +0xc78

goroutine 19 [select]:
net/http.(*persistConn).writeLoop(0xc8200fda20)
	/goroot/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/goroot/src/net/http/transport.go:686 +0xc9d
```

That seems like a show-stopper. How did this ship in 0.3.1?

**Comment by hobbeswalsh** [comment]
Oh, right, this is the x86_64 linux binary.

**Comment by jamesob** [comment]
+1

**Comment by jefferai** [comment]
Is this new behavior? Is this something that worked for you in 0.3.0?

**Comment by jefferai** [comment]
Specifically:

Did it work in 0.2?
Did it work in 0.3?

None of the ldap code changed in 0.3.1; neither did the auth code. So the answer to how it shipped in 0.3.1 is "we haven't had a report of this problem before and none of the relevant code changed".

**Comment by hobbeswalsh** [comment]
@jefferai this appears to be a 0.3.1 regression.

```
$ ./0.2.0/vault auth -method=ldap username="My Username"
Password (will be hidden):
Successfully authenticated! The policies that are associated
with this token are listed below:

root
```

```
$ ./0.3.0/vault auth -method=ldap username="My Username"
Password (will be hidden):
Successfully authenticated!
The policies that are associated with this token
arelisted below:

root

```
$ ./0.3.1/vault auth -method=ldap username="My Username"
Password (will be hidden):
panic: interface conversion: interface is nil, not float64

goroutine 1 [running]:
github.com/hashicorp/vault/command.(*AuthCommand).Run(0xc8200fd8c0, 0xc82000a1b0, 0x1, 0x1, 0xc82013c090)
	/gopath/src/github.com/hashicorp/vault/command/auth.go:206 +0x2041
github.com/mitchellh/cli.(*CLI).Run(0xc8200fd810, 0xc8200fd810, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/Godeps/_workspace/src/github.com/mitchellh/cli/cli.go:112 +0x655
github.com/hashicorp/vault/cli.RunCustom(0xc82000a190, 0x3, 0x3, 0xc820122720, 0x2)
	/gopath/src/github.com/hashicorp/vault/cli/main.go:44 +0x452
github.com/hashicorp/vault/cli.Run(0xc82000a190, 0x3, 0x3, 0x989300)
	/gopath/src/github.com/hashicorp/vault/cli/main.go:11 +0x4c
main.main()
	/gopath/src/github.com/hashicorp/vault/main.go:10 +0x60

goroutine 5 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 7 [semacquire]:
sync.runtime_Syncsemacquire(0xc820140040)
	/usr/local/go/src/runtime/sema.go:237 +0x201
sync.(*Cond).Wait(0xc820140030)
	/usr/local/go/src/sync/cond.go:62 +0x9b
io.(*pipe).read(0xc820140000, 0xc820162000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:52 +0x2d2
io.(*PipeReader).Read(0xc820028168, 0xc820162000, 0x1000, 0x1000, 0xc82001a418, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:134 +0x50
bufio.(*Scanner).Scan(0xc820016800, 0x2)
	/usr/local/go/src/bufio/scan.go:180 +0x877
github.com/hashicorp/vault/command.(*Meta).FlagSet.func1(0xc820016800, 0xc8200fd8c0)
	/gopath/src/github.com/hashicorp/vault/command/meta.go:239 +0x25
created by github.com/hashicorp/vault/command.(*Meta).FlagSet
	/gopath/src/github.com/hashicorp/vault/command/meta.go:242 +0x5e6

goroutine 10 [select, locked to thread]:
runtime.gopark(0x1380530, 0xc820027728, 0x1121188, 0x6, 0x42c518, 0x2)
	/usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc820027728, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc820027728)
	/usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 18 [IO wait]:
net.runtime_pollWait(0x2ab6609ad778, 0x72, 0xc82000e160)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc820127e20, 0x72, 0x0, 0x0)
	/goroot/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820127e20, 0x0, 0x0)
	/goroot/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820127dc0, 0xc8200d1400, 0x400, 0x400, 0x0, 0x2ab6609a7050, 0xc82000e160)
	/goroot/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820028238, 0xc8200d1400, 0x400, 0x400, 0x0, 0x0, 0x0)
	/goroot/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc82018a270, 0x2ab6609adbc8, 0xc820028238, 0x5, 0x0, 0x0)
	/goroot/src/crypto/tls/conn.go:455 +0xcc
crypto/tls.(*Conn).readRecord(0xc82007b340, 0x1380517, 0x0, 0x0)
	/goroot/src/crypto/tls/conn.go:540 +0x2d1
crypto/tls.(*Conn).Read(0xc82007b340, 0xc820185000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/goroot/src/crypto/tls/conn.go:901 +0x167
net/http.noteEOFReader.Read(0x2ab6609b6598, 0xc82007b340, 0xc8200fdb28, 0xc820185000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/goroot/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc820480780, 0xc820185000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc8203cf980)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8203cf980, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8200fdad0)
	/goroot/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/goroot/src/net/http/transport.go:685 +0xc78

goroutine 19 [select]:
net/http.(*persistConn).writeLoop(0xc8200fdad0)
	/goroot/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/goroot/src/net/http/transport.go:686 +0xc9d
```

**Comment by jefferai** [comment]
My guess is that this is somehow related to fixing the builds to be static instead of dynamic. I can send you a dynamically-linked build for testing if you like.

**Comment by jefferai** [comment]
Alternately, can you self-compile for testing?

**Comment by hobbeswalsh** [comment]
yeah, you mean using `go get`? I can try that.

**Comment by jefferai** [comment]
Try two things, if you don't mind:

1) The linux_x86 (32-bit) version
2) A self-compiled version. The ideal way to do this to mimic how we build would be to clone the repo, run `godep restore` to get the same versions of dependencies, and then run "make dev" to build. The binary will be output to the bin/ directory. Once you can do that successfully I may give you a couple variations to try. Can you also let me know what `file bin/vault` says?

**Comment by jefferai** [comment]
Another question -- are you using both a 0.3.1 client and 0.3.1 server?

**Comment by hobbeswalsh** [comment]
ugh, okay -- 

```
go/test/syntax/import.go:10:6: expected ';', found ',' (and 1 more errors)
can't load package: package ./go/test/bench/shootout: C source files not allowed when not using cgo or SWIG: binary-tree.c chameneosredux.c fannkuch.c fasta.c k-nucleotide.c mandelbrot.c meteor-contest.c nbody.c pidigits.c regex-dna.c reverse-complement.c spectral-norm.c threadring.c
make: *** [generate] Error 1
```

**Comment by hobbeswalsh** [comment]
do you have a Vagrant box you can test this on? meanwhile I'll try to get it to work.

**Comment by jefferai** [comment]
@hobbeswalsh Did you run `make dev` or `make`?

**Comment by jefferai** [comment]
Also it looks like you're trying to build Go there, not trying to build Vault.

**Comment by hobbeswalsh** [comment]
yeah, was trying to do two things at once and unzipped go distribution on top of vault. let me try again.

**Comment by jefferai** [comment]
OK. Are you using the same server/client versions?

**Comment by hobbeswalsh** [comment]
server is 0.2.0, clients 0.2.0 and 0.3.0 work. 0.3.1 does not.

I can't upgrade the server right now, as that would cause downtime and we'd need to re-unseal.

**Comment by hobbeswalsh** [comment]
I just ran the version of Vault built with `make dev` and it failed with the same stack trace.

**Comment by jefferai** [comment]
I'm surprised that 0.3.0 works, as the output formats and CLI code haven't changed between 0.3.0 and 0.3.1, but generally speaking 0.2.0 and 0.3.x are *not* compatible client and server versions.

When you get downtime, please try using a same version of the server and client.

Also, based on the stack trace, this is almost certainly only affecting the CLI; the HTTP API should be fine.

**Comment by hobbeswalsh** [comment]
FWIW this is also broken on the i686 platform.

**Comment by hobbeswalsh** [comment]
I guess for now the guidance will be to use version 0.2.0 of Vault -- but what's the upgrade path for this, then? I upgrade the server and then tell everyone at the company: "sorry, your CLI might just stop working all of a sudden. Please get a new version." Not the most user-friendly upgrade path, but I understand that there are some potentially-breaking changes between 0.2 and 0.3

**Comment by jefferai** [comment]
The upgrade path, unless explicitly specified otherwise, is to assume that Vault's API can change between minor versions of Vault (e.g. 0.2 and 0.3). The Vault CLI is simply an API client.

We try to ensure that any breaking changes are specified (and a number of them are noted in the 0.3 changelog entry).

**Comment by hobbeswalsh** [comment]
You can close this if you'd like. I have all the information I need. Thanks.

**Comment by jefferai** [comment]
OK, thanks!
